Docs 9377 encode decode duke (#1702)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/dataweave-cookbook-decode-encode.adoc

= Decode and Encode Base64
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:keywords: studio, anypoint, transform, transformer, format, xml, metadata, dataweave, data weave, datamapper, dwl, dfl, dw, output structure, input structure, map, mapping

The DataWeave examples show how to convert a file stream into Base64 and to convert a Base64 string into a file stream. They use a PDF file as the input and output.
//LINK TO DW 1.0 EXAMPLES:
include::partial$dataweave1-links.adoc[tag=dataweave1Examples]

[[xml_config]]

> == XML Configuration Example

Reviewer: Reading this first made me confused about whether the example was done and what the other sections were about. I'd move this to the bottom as a full Mule app example featuring the functionality in the previous examples that you can try in Studio.

Author: moved to end

. Copy and paste the following example between the `<mule/>` elements of a Mule flow in Anypoint Studio.
+
[source,XML,linenums]
----
<http:listener-config name="HTTP_Listener_config"
                      doc:name="HTTP Listener config" >
    <http:listener-connection
                            host="0.0.0.0"
                            port="8081" />
</http:listener-config>
<file:config name="File_Config"
             doc:name="File Config" >
    <file:connection workingDir="/Users/me/Downloads" />
</file:config>
<flow name="encode-decode-base64">
    <http:listener doc:name="Listener"
                 path="/transform"
                 config-ref="HTTP_Listener_config"/>
    <file:read doc:name="Read"
             config-ref="File_Config"
             path="pdf-test.pdf" />
    <ee:transform doc:name="Transform Message">
        <ee:message>
          <ee:set-payload><![CDATA[%dw 2.0
import * from dw::core::Binaries
output text/plain
---
toBase64(payload as Binary)]]></ee:set-payload>
        </ee:message>
    </ee:transform>
    <ee:transform doc:name="Transform Message" >
        <ee:message >
          <ee:set-payload ><![CDATA[%dw 2.0
import * from dw::core::Binaries

> output application/octet-stream

Reviewer: If these are docs for 4.3 exclusively, we could use `output binary` which is more concise and easier to remember, with a note perhaps saying its equivalent to writing that mime type.

Reviewer: Actually, I think this would be the correct use for custom mime type using `output binary with application/pdf`.

Author: cool, `output binary` works in Studio using 4.3.0. And when ordered `output application/pdf with binary`, it also worked. Updated to the latter.

* In the Write operation, change the *Path* value `/Users/me/pdf-test.pdf"` to the location and file name that you want to output the file.
. Run the project in Studio.
. Load the `0.0.0.0:8081/transform` into a browser.
+
This action creates a PDF file in the specified location and generates a separate `transform` file with the message in the Set Payload component, `PDF file created in the directory specified in the Write operation.`

[[ex01]]
== Encode a PDF File to Base64

You can simulate the transformation in the Read operation from the <<xml_config, XML configuration>>. This example performs Base64 encoding on a PDF file. More specifically, it converts a Base64 file stream (`application/octet-stream`) to Binary and outputs the Binary in the `text/plain` format.

The example uses the following functions:

*  `readUrl` from the `Core (dw::Core)` module to input a PDF file.
+
The example assumes that the PDF file is in the `src/main/resources` directory of a Mule project in Anypoint Studio. It provides the name of the PDF file and specifies the MIME type as `application/octet-stream`.

> * `fromBase64` from the `dw::core::Binaries` module to transform a Base64 string to a binary value.

Reviewer: Shouldn't this go beneath the other item?

Author: done

//NOTE that we can't yet automate this example. Outputs text format.
//Need to check if PDF input is supported by validator.
.DataWeave Script:
[source,dataweave,linenums]
----
%dw 2.0
import * from dw::core::Binaries
var myPDF = readUrl("classpath://pdf-test.pdf", "application/octet-stream")
output text/plain
---
toBase64(myPDF as Binary)
----

.Output (partial results):
[source,json,linenums]
----
JVBERi0xLjYNJeLjz9MNCjM3IDAgb2JqIDw8L0xpbmVhcml6ZWQgMS9MIDIwNTk3L08gNDAvRSAx
...
----

[[ex02]]
== Decode Base64

You can simulate the transformation in the Write operation from the <<xml_config, XML configuration>>. This DataWeave example converts a Base64 file stream (`application/octet-stream`) to Binary and outputs the Binary in the `text/plain` format. It does not generate a PDF file; a Write operation is required to generate such a file.

The examples use the following functions from the `dw::core::Binaries` module:

*  `readUrl` to input a PDF file.
+
The example assumes that the PDF file is in the `src/main/resources` directory of a Mule project in Anypoint Studio. It provides the name of the PDF file and specifies the MIME type as `application/octet-stream`.

> * `fromBase64` in the `dw::core::Binaries` module to transform a Base64 string to a binary value.

Reviewer: Same as previous comment.

Author: done

The script reads the input PDF as an octet-stream MIME type. From within the call to `fromBase64`, the script coerces that input to a string value before coercing the string to Base64.

//NOTE that we can't yet automate this example. Outputs text format.
//Need to check if PDF input is supported by validator.
.DataWeave Script:
[source,dataweave,linenums]
----
%dw 2.0
import * from dw::core::Binaries
var myPDF = readUrl("classpath://pdf-test.pdf", "application/octet-stream")
var myPDFasBinary = toBase64(myPDF as Binary)

> output text/plain

Reviewer: This is confusing, shouldn't it be binary too as in the other example? In a mule app this would fail

Author: That seemed to work in Studio (no error thrown). `readUrl` doesn't accept `"application/pdf with binary"` as the second arg but is ok with `"binary"`.

Author: Modified the `fromBase64` example with the suggested change, replacing `output text/plain` with `output application/pdf with binary`

Author: I expected that, too, but I got those results when I ran this in the Transform Message component:
```
%dw 2.0
import * from dw::core::Binaries
var myPDF = readUrl("classpath://pdf-test.pdf", "application/octet-stream")
//var myPDFasBinary = toBase64(myPDF as Binary)
output application/PDF with binary
//output application/octet-stream
---
fromBase64(payload as String) as Binary
```

I put in a logger (with fx set to `"testing output" ++ payload`)  after that TM component, saved, stopped, and restarted the app, and got this result  in the console:
```
testing output%PDF-1.6
%����
37 0 obj <</Linearized 1/L 20597/O 40/E 14115/N 1/T 19795/H [ 1005 215]>>
endobj
                 
xref
37 34
0000000016 00000 n
0000001386 00000 n
```
same went for the DW script in a standalone TM component (not in that flow).
Script:
```
%dw 2.0
import * from dw::core::Binaries
var myPDF = readUrl("classpath://pdf-test.pdf", "binary")
var myPDFasBinary = toBase64(myPDF as Binary)
output application/PDF with binary
---
fromBase64(myPDFasBinary as String) as Binary
```
returned

```
%PDF-1.6
%����
37 0 obj <</Linearized 1/L 20597/O 40/E 14115/N 1/T 19795/H [ 1005 215]>>
endobj
                 
xref
37 34
0000000016 00000 n
```

---
fromBase64(myPDFasBinary as String) as Binary
----

.Output (partial results):
[source,json,linenums]
----

> %PDF-1.6

Reviewer: Maybe clarify this is the text representation of a pdf?

Author: done

Reviewer: I think if you try this now it will output gibberish because it's now binary. The output needs updating and the note now has to mention it's the binary representation.

Author: I expected that, too, but I got those results when I ran this in the Transform Message component:
```
%dw 2.0
import * from dw::core::Binaries
//var myPDF = readUrl("classpath://pdf-test.pdf", "application/octet-stream")
//var myPDFasBinary = toBase64(myPDF as Binary)
output application/PDF with binary
//output application/octet-stream
---
fromBase64(payload as String) as Binary
```

I put in a logger (with fx set to `"testing output" ++ payload`)  after that TM component, saved, stopped, and restarted the app, and got this result  in the console:
```
testing output%PDF-1.6
%����
37 0 obj <</Linearized 1/L 20597/O 40/E 14115/N 1/T 19795/H [ 1005 215]>>
endobj
                 
xref
37 34
0000000016 00000 n
0000001386 00000 n
```
same went for the DW script in a standalone TM component (not in that flow).
Script:
```
%dw 2.0
import * from dw::core::Binaries
var myPDF = readUrl("classpath://pdf-test.pdf", "binary")
var myPDFasBinary = toBase64(myPDF as Binary)
output application/PDF with binary
---
fromBase64(myPDFasBinary as String) as Binary
```
returned

```
%PDF-1.6
%����
37 0 obj <</Linearized 1/L 20597/O 40/E 14115/N 1/T 19795/H [ 1005 215]>>
endobj
                 
xref
37 34
0000000016 00000 n
```

Author: those are partial results but you can see `%����` bit

= Decode and Encode Base64
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:keywords: studio, anypoint, transform, transformer, format, xml, metadata, dataweave, data weave, datamapper, dwl, dfl, dw, output structure, input structure, map, mapping

> The DataWeave examples show how to convert a file stream into Base64 and to convert a Base64 string into a file stream. They use a PDF file as the input and output.

Reviewer 2: ```suggestion
The following DataWeave examples show how to convert a file stream into Base64 and to convert a Base64 string into a file stream. They use a PDF file as the input and output.
```

Author: done

//LINK TO DW 1.0 EXAMPLES:
include::partial$dataweave1-links.adoc[tag=dataweave1Examples]

[[ex01]]
== Encode a PDF File to Base64

This example performs Base64 encoding on a PDF file. It simulates the transformation in the Read operation from the <<xml_config, XML configuration>> by converting a Base64 file stream (`application/octet-stream`) to Binary and outputting the Binary in the `text/plain` format. 

The example uses the following functions:

* `fromBase64` from the `dw::core::Binaries` module to transform a Base64 string to a binary value.
*  `readUrl` from the `Core (dw::Core)` module to input a PDF file.
+

> The example assumes that the PDF file is in the `src/main/resources` directory of a Mule project in Anypoint Studio. It provides the name of the PDF file and specifies the MIME type as `"application/octet-stream"`. Using `"binary"` as the second `readUrl` argument is also valid for this example.

Reviewer 2: ```suggestion
The example assumes that the PDF file is in the `src/main/resources` directory of a Mule project in Anypoint Studio. It provides the name of the PDF file and specifies the MIME type as `application/octet-stream`. Using `binary` as the second `readUrl` argument is also valid for this example.
```

Author: the arguments are surrounded by quotation marks, so I needed to put quotes around them

//NOTE that we can't yet automate this example. Outputs text format.
//Need to check if PDF input is supported by validator.
.DataWeave Script:
[source,dataweave,linenums]
----
%dw 2.0
import * from dw::core::Binaries
var myPDF = readUrl("classpath://pdf-test.pdf", "application/octet-stream")
output text/plain
---
toBase64(myPDF as Binary)
----

.Output (partial results):
[source,json,linenums]
----
JVBERi0xLjYNJeLjz9MNCjM3IDAgb2JqIDw8L0xpbmVhcml6ZWQgMS9MIDIwNTk3L08gNDAvRSAx
...
----

[[ex02]]
== Decode Base64

This example simulates the transformation in the Write operation from the <<xml_config, XML configuration>> by converting a Base64 file stream (`application/octet-stream`) to Binary and outputting the Binary in the `application/PDF with binary` format. The example does not generate a PDF file; a Write operation is required to generate such a file. 

> The script reads the input PDF as an octet-stream MIME type. From within the call to `fromBase64`, the script coerces that input to a string value before coercing the string to Base64.
> 
> The examples use the following functions:
> 
> * `fromBase64` in the `dw::core::Binaries` module to transform a Base64 string to a binary value.
> * `readUrl` to input a PDF file.
> +
> The example assumes that the PDF file is in the `src/main/resources` directory of a Mule project in Anypoint Studio. It provides the name of the PDF file and specifies the MIME type as `application/octet-stream`. Using the MIME type `"binary"` as the second `readUrl` argument is also valid for this example.

Reviewer 2: ```suggestion
The examples use the following functions:

* `fromBase64` in the `dw::core::Binaries` module to transform a Base64 string to a binary value.
* `readUrl` to input a PDF file.
+
The script reads the input PDF as an octet-stream MIME type. From within the call to `fromBase64`, the script coerces that input to a string value before coercing the string to Base64.
The example assumes that the PDF file is in the `src/main/resources` directory of a Mule project in Anypoint Studio. It provides the name of the PDF file and specifies the MIME type as `application/octet-stream`. Using the MIME type `"binary"` as the second `readUrl` argument is also valid for this example.
```

Reviewer 2: I would move the sentence afterwards the introduction of the used functions, because the sentence mentions specifically the `fromBase64` function.

Author: done, removed duplicative language

//NOTE that we can't yet automate this example. Outputs text format.
//Need to check if PDF input is supported by validator.
.DataWeave Script:
[source,dataweave,linenums]
----
%dw 2.0
import * from dw::core::Binaries
var myPDF = readUrl("classpath://pdf-test.pdf", "application/octet-stream")
var myPDFasBinary = toBase64(myPDF as Binary)
output application/PDF with binary
---
fromBase64(myPDFasBinary as String) as Binary
----

The output is the text representation of the PDF. 

.Output (partial results):
[source,json,linenums]
----
%PDF-1.6
%����
37 0 obj <</Linearized 1/L 20597/O 40/E 14115/N 1/T 19795/H [ 1005 215]>>
endobj
                 
xref
37 34
0000000016 00000 n
...
----

[[xml_config]]
== XML Configuration Example

> In Anypoint Studio, you can create and run a Mule flow that encodes and decodes a PDF file. After you complete the procedure, the example will read the file from the specified local directory using the Read operation from the File module. It will write the PDF to the specified directory by using the Write operation in the File module. It will use Transform Message components to encode the file to Base64 and to decode that output from Base64.

Reviewer 2: ```suggestion
In Anypoint Studio, you can create and run a Mule flow that encodes and decodes a PDF file. After you complete the procedure, the example reads the file from the specified local directory using the File Connector Read operation. It writes the PDF to the specified directory by using the File Connector Write operation, and uses Transform Message components to encode the file to Base64 and to decode that output from Base64.
```

Author: done

> . Create a Mule project in Anypoint Studio.
> . Use *Add Module* in the Mule Palette to add the File module.
> . Copy and paste the following example between the `<mule/>` elements of a Mule flow in Anypoint Studio.

Reviewer 2: ```suggestion
To create the Mule flow:

. In Studio, select *File > New > Mule Project*.
. Enter a name for your Mule project and click *Finish*.
. In the *Mule Palette* view, click *(X) Search in Exchange*.
. In the *Add Dependencies to Project* window, type `file` in the search field.
. Click *File Connector* in Available modules.
. Click *Add*.
. Click *Finish*.
. In the Studio canvas, navigate to the *Configuration XML* tab.
. Copy and paste the following XML example between the `<mule/>` elements of the flow:
```

Author: done mostly as substeps, keeping high-level steps in place.

             config-ref="File_Config"
             path="pdf-test.pdf" />
    <ee:transform doc:name="Transform Message">
        <ee:message>
          <ee:set-payload><![CDATA[%dw 2.0
import * from dw::core::Binaries
output text/plain
---
toBase64(payload as Binary)]]></ee:set-payload>
        </ee:message>
    </ee:transform>
    <ee:transform doc:name="Transform Message" >
        <ee:message >
          <ee:set-payload ><![CDATA[%dw 2.0
import * from dw::core::Binaries
output application/pdf with binary
---
fromBase64(payload as String) as Binary]]></ee:set-payload>
        </ee:message>
    </ee:transform>
    <file:write doc:name="Write"
              path="/Users/me/pdf-test-result.pdf"/>
    <set-payload
              value='#["PDF file created in the directory specified in the Write operation."]'
              doc:name="Set Payload" />
</flow>
----
+
Note that it is valid to replace the directive `output application/pdf with binary` in the second Transform Message component with `output application/octet-stream`.
+
. Place a PDF file of your choice into a local directory on your machine.
. Correct the input and output directories of the Read and Write operations:
+

> * In the Read operation, change the *File Path* value `/Users/me/Downloads" to the location of your PDF file.
> * In the Write operation, change the *Path* value `/Users/me/pdf-test.pdf"` to the location and file name that you want to output the file.

Reviewer 2: ```suggestion
* In the *Read* operation, change the *File Path* value `/Users/me/Downloads" to the location of your PDF file.
* In the *Write* operation, change the *Path* value `/Users/me/pdf-test.pdf"` to the location and file name that you want to output the file.
```

Reviewer 2: In Connectors we style the operation name in bold, just adding it here for consistency.

Author: done, thx for heads up
